src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1321,21 +1386,14 @@ public SolutionCompilationState WithoutFrozenSourceGeneratedDocuments()
     /// generated file open, we need to make sure everything lines up.
     /// </summary>
     public SolutionCompilationState WithFrozenSourceGeneratedDocuments(
-        ImmutableArray<(SourceGeneratedDocumentIdentity documentIdentity, DateTime generationDateTime, SourceText sourceText)> documents)
+        ImmutableArray<(SourceGeneratedDocumentIdentity documentIdentity, DateTime generationDateTime, SourceText sourceText, SyntaxNode? syntaxNode)> documents)
     {
-        // We won't support freezing multiple source generated documents more than once in a chain, simply because we have no need
-        // to support that; these solutions are created on demand when we need to operate on an open source generated document,
-        // and so those are always forks off the main solution. There's also a bit of a design question -- does calling this a second time
-        // leave the existing frozen documents in place, or replace them? It depends on the need, but until then we'll cross that bridge
-        // if/when we need it.
-        Contract.ThrowIfFalse(FrozenSourceGeneratedDocumentStates == null, $"We shouldn't be calling {nameof(WithFrozenSourceGeneratedDocuments)} on a solution with frozen source generated documents.");
-
         if (documents.IsEmpty)
             return this;
 
         // We'll keep track if every document we're reusing is the exact same as the final generated output we already have
-        using var _ = ArrayBuilder<SourceGeneratedDocumentState>.GetInstance(documents.Length, out var documentStates);
-        foreach (var (documentIdentity, generationDateTime, sourceText) in documents)
+        using var _ = PooledDictionary<DocumentId, SourceGeneratedDocumentState>.GetInstance(out var documentStates);
```

Comment:
Rather than having a separate dictionary created here, it seems we could just reuse our existing dictionary object (or fill in an empty one if still null.) Then the loop calling TryAdd goes away, and the calls to .Add on this just become a call to SetItem on the dictionary. That way the number of updates is more in line with the number of documents being passed here, as opposed to us copying around the stuff that _isn't_ changing. The check of "// If every document we looked at matched what we've already generated, we have nothing new to do" just becomes an identity check at that point.

Reply:
Not sure what you mean by "existing dictionary". Do you mean `FrozenSourceGeneratedDocumentStates`? If so, thats not quite a dictionary. eg, looks like calling `SetState` won't add, it will throw if the id doesn't exist. Will leave this for later